src/qttools/datastructures/dsbcsr.py

```diff
@@ -340,7 +344,77 @@ def __neg__(self) -> "DSBCSR":
+            raise ValueError("Block sizes do not match.")
+        stack_indices = xp.ndindex(self.data.shape[:-1])
+        product_rows, product_cols = product_sparsity_pattern(
+            sparse.coo_matrix(
+                (xp.ones(self.nnz), (self.spy())), shape=self.shape[-2:]
+            ),
+            sparse.coo_matrix(
+                (xp.ones(other.nnz), (other.spy())), shape=other.shape[-2:]
+            ),
+        )
+        block_sort_index = compute_block_sort_index(
+            product_rows, product_cols, self.block_sizes
+        )
+        product = DSBCSR(
+            data=xp.zeros(self.stack_shape + (product_rows.size,), dtype=self.dtype),
+            cols=product_cols[block_sort_index],
+            rowptr_map=compute_ptr_map(product_rows, product_cols, self.block_sizes),
+            block_sizes=self.block_sizes,
+            global_stack_shape=self.global_stack_shape,
+        )
+        for stack_index in stack_indices:
+            temp_product = sparse.csr_matrix(
+                (self.data[stack_index], (self.spy())), shape=self.shape[-2:]
+            ) @ sparse.csr_matrix(
+                (other.data[stack_index], (other.spy())), shape=other.shape[-2:]
+            )
+            product.data[stack_index, :] = temp_product[product.spy()]
+        return product
+
+    @DSBSparse.block_sizes.setter
+    def block_sizes(self, block_sizes: ArrayLike) -> None:
+        """Sets new block sizes for the matrix.
+        Parameters
+        ----------
+        block_sizes : array_like
+            The new block sizes.
+        """
+        if self.distribution_state == "nnz":
+            raise NotImplementedError(
+                "Cannot reassign block-sizes when distributed through nnz."
+            )
+        if sum(block_sizes) != self.shape[-1]:
+            raise ValueError("Block sizes do not match matrix shape.")
+        rows, cols = self.spy()
+        # Compute canonical ordering of the matrix.
+        inds_bcsr2canonical = xp.lexsort(xp.vstack((cols, rows)))
+        canonical_rows = rows[inds_bcsr2canonical]
+        canonical_cols = cols[inds_bcsr2canonical]
+        # Compute the index for sorting by the new block-sizes.
+        inds_canonical2bcsr = compute_block_sort_index(
+            canonical_rows, canonical_cols, block_sizes
+        )
+        # Mapping directly from original block-ordering to the new
+        # block-ordering is achieved by chaining the two mappings.
+        inds_bcsr2bcsr = inds_bcsr2canonical[inds_canonical2bcsr]
+        self.data[:] = self.data[..., inds_bcsr2bcsr]
+        self.cols = self.cols[inds_bcsr2bcsr]
+        # Compute the rowptr map for the new block-sizes.
+        self.rowptr_map = compute_ptr_map(canonical_rows, canonical_cols, block_sizes)
+        self._block_sizes = xp.asarray(block_sizes, dtype=int)
+        self._block_offsets = xp.hstack(([0], xp.cumsum(block_sizes)))
+        self.num_blocks = len(block_sizes)
```

Comment:
You'll also need to invalidate the block_slice_cache here, so add `self._block_slice_cache = {}` at the end (also in the COO)